File: src/memory.ml
external set_memory :
  Module.t ->
  int ->
  int ->
  string ->
  bytes list ->
  bool list ->
  Expression.t list ->
  int list ->
  bool ->
  bool ->
  string ->
  unit = "caml_binaryen_set_memory__bytecode" "caml_binaryen_set_memory"

type segment = { data : bytes; kind : segment_kind; size : int }
and segment_kind = Passive | Active of { offset : Expression.t }

(** Module, initial size, maximum size, export name, segments, shared, memory64, moduleName . *)
let set_memory wasm_mod initial maximum export_name (segments : segment list)
    shared memory64 moduleName =
  let split_segments segments =
    List.fold_right
      (fun { data; kind; size }
           (segment_data, segment_passive, segment_offsets, segment_sizes) ->
        match kind with
        | Active { offset } ->
            ( data :: segment_data,
              false :: segment_passive,
              offset :: segment_offsets,
              size :: segment_sizes )
        | Passive ->
            ( data :: segment_data,
              true :: segment_passive,
              Expression.Null.make () :: segment_offsets,
              size :: segment_sizes ))
      segments ([], [], [], [])
  in
  let segment_data, segment_passive, segment_offsets, segment_sizes =
    split_segments segments
  in
  set_memory wasm_mod initial maximum export_name segment_data segment_passive
    segment_offsets segment_sizes shared memory64 moduleName

external has_memory : Module.t -> bool = "caml_binaryen_has_memory"

external get_initial : Module.t -> string -> int
  = "caml_binaryen_memory_get_initial"

external has_max : Module.t -> string -> bool = "caml_binaryen_memory_has_max"
external get_max : Module.t -> string -> int = "caml_binaryen_memory_get_max"

external is_shared : Module.t -> string -> bool
  = "caml_binaryen_memory_is_shared"

external is_64 : Module.t -> string -> bool = "caml_binaryen_memory_is_64"

let unlimited = -1

external get_num_segments : Module.t -> int
  = "caml_binaryen_get_num_memory_segments"

external get_segment_byte_offset : Module.t -> int -> int
  = "caml_binaryen_get_memory_segment_byte_offset"

external get_segment_passive : Module.t -> int -> bool
  = "caml_binaryen_get_memory_segment_passive"

external get_segment_data : Module.t -> int -> bytes
  = "caml_binaryen_get_memory_segment_data"

File: src/memory.js
//Provides: caml_binaryen_set_memory
//Requires: caml_list_to_js_array, caml_convert_bytes_to_array
//Requires: caml_jsstring_of_string, caml_js_from_bool
function caml_binaryen_set_memory(
  wasm_mod,
  initial,
  maximum,
  exportName,
  segmentNames,
  segmentData,
  segmentPassives,
  segmentOffsets,
  segmentSizes,
  shared,
  memory64,
  memoryName
) {
  var datas = caml_list_to_js_array(segmentData);
  var passives = caml_list_to_js_array(segmentPassives);
  var offsets = caml_list_to_js_array(segmentOffsets);
  var segs = caml_list_to_js_array(segmentNames).map(function (name, idx) {
    return {
      name: caml_jsstring_of_string(name),
      data: caml_convert_bytes_to_array(datas[idx]),
      passive: caml_js_from_bool(passives[idx]),
      offset: offsets[idx],
    };
  });

  return wasm_mod.setMemory(
    initial,
    maximum,
    caml_jsstring_of_string(exportName),
    segs,
    caml_js_from_bool(shared),
    caml_js_from_bool(memory64),
    caml_jsstring_of_string(memoryName)
  );
}
//Provides: caml_binaryen_set_memory__bytecode
//Requires: caml_binaryen_set_memory
function caml_binaryen_set_memory__bytecode() {
  return caml_binaryen_set_memory(
    arguments[0],
    arguments[1],
    arguments[2],
    arguments[3],
    arguments[4],
    arguments[5],
    arguments[6],
    arguments[7],
    arguments[8],
    arguments[9],
    arguments[10],
    arguments[11]
  );
}

//Provides: caml_binaryen_has_memory
//Requires: caml_js_to_bool
function caml_binaryen_has_memory(mod) {
  return caml_js_to_bool(mod.hasMemory());
}

//Provides: caml_binaryen_memory_get_initial
//Requires: caml_jsstring_of_string
function caml_binaryen_memory_get_initial(mod, memoryName) {
  var memory_info = mod.getMemoryInfo(caml_jsstring_of_string(memoryName));
  return memory_info.initial;
}

//Provides: caml_binaryen_memory_has_max
//Requires: caml_js_to_bool, caml_jsstring_of_string
function caml_binaryen_memory_has_max(mod, memoryName) {
  var memory_info = mod.getMemoryInfo(caml_jsstring_of_string(memoryName));
  if (memory_info.max != null) {
    return caml_js_to_bool(true);
  } else {
    return caml_js_to_bool(false);
  }
}

//Provides: caml_binaryen_memory_get_max
//Requires: caml_jsstring_of_string
function caml_binaryen_memory_get_max(mod, memoryName) {
  var memory_info = mod.getMemoryInfo(caml_jsstring_of_string(memoryName));
  if (memory_info.max != null) {
    return memory_info.max;
  } else {
    // This ensures that our return is equal to Memory.unlimited
    return -1;
  }
}

//Provides: caml_binaryen_memory_is_shared
//Requires: caml_js_to_bool, caml_jsstring_of_string
function caml_binaryen_memory_is_shared(mod, memoryName) {
  var memory_info = mod.getMemoryInfo(caml_jsstring_of_string(memoryName));
  return caml_js_to_bool(memory_info.shared);
}

//Provides: caml_binaryen_memory_is_64
//Requires: caml_js_to_bool, caml_jsstring_of_string
function caml_binaryen_memory_is_64(mod, memoryName) {
  var memory_info = mod.getMemoryInfo(caml_jsstring_of_string(memoryName));
  return caml_js_to_bool(memory_info.is64);
}

//Provides: caml_binaryen_get_num_memory_segments
function caml_binaryen_get_num_memory_segments(wasm_mod) {
  return wasm_mod.getNumMemorySegments();
}

//Provides: caml_binaryen_get_memory_segment_byte_offset
//Requires: caml_jsstring_of_string
function caml_binaryen_get_memory_segment_byte_offset(wasm_mod, name) {
  var info = wasm_mod.getMemorySegmentInfo(caml_jsstring_of_string(name));
  return info.offset;
}

//Provides: caml_binaryen_get_memory_segment_passive
//Requires: caml_jsstring_of_string
function caml_binaryen_get_memory_segment_passive(wasm_mod, name) {
  var info = wasm_mod.getMemorySegmentInfo(caml_jsstring_of_string(name));
  return info.passive;
}

//Provides: caml_binaryen_get_memory_segment_data
//Requires: caml_bytes_of_array, caml_jsstring_of_string
function caml_binaryen_get_memory_segment_data(wasm_mod, name) {
  var info = wasm_mod.getMemorySegmentInfo(caml_jsstring_of_string(name));
  return caml_bytes_of_array(info.data);
}
